src/fparser/two/Fortran2008.py

```diff
@@ -96,6 +101,241 @@ class Program_Unit(Program_Unit_2003):  # R202
     subclass_names.append("Submodule")
 
 
+class Component_Attr_Spec(STRINGBase):  # R437
```

Comment:
Wouldn't that still require re-implementing the match-method in the F08 class? Since `match` is a staticmethod it would require to refer explicitly to the class variable (e.g., `Component_Attr_Spec.attributes`) and thus overwriting it in a subclass would not change the behaviour in the match method.

I was able to get it to work by changing `match` to a classmethod in the F03 class and use `cls.attributes` in the matching, i.e.,
```python
class Component_Attr_Spec(STRINGBase):
    ...
    attributes = [...]
    @classmethod
    def match(cls, string):
        return STRINGBase.match(cls.attributes, string)
```
Then I can setup the F08 class as
```python
class Component_Attr_Spec(Component_Attr_Spec_2003):
    subclass_names = Component_Attr_Spec_2003[:]
    subclass_names.append('Codimension_Attr_Spec')
    Component_Attr_Spec_2003.attributes.append('CONTIGUOUS')
    match = Component_Attr_Spec_2003.match
```
Interestingly, it doesn't work if I extend `attributes` in the same way as `subclass_names`. This does somewhat exceed my knowledge about edge cases of Python inheritance...

Reply:
OK, there's good and bad news:

After some discussion and digging with @mlange05 we identified the cause for that behaviour. In [utils.py:323](https://github.com/stfc/fparser/blob/master/src/fparser/two/utils.py#L323) the `match` method is obtained by looking into the `cls.__dict__`:
``` match = cls.__dict__.get('match')```
instead of using `getattr`
```match = getattr(cls, 'match') if hasattr(cls, 'match') else None```
which breaks native inheritance. Thus, without explicitly implementing the match method in the subclass (which above assignment does), there is no 'match' in `__dict__`. And with the assignment, `cls` is not the subclass but the parent. (Interesting Python feature BTW).

With that said, I can extend `attributes` in the same way as `subclass_names` and leave out the definition of `match` if the F03 match is turned into a classmethod and utils.py is altered in the way above.

The bad news: this change breaks other tests (src/fparser/two/tests/fortran2003/test_defined_operators_r703_r723.py).

The good news: my knowledge about polymorphism in Python wasn't so far off after all... ;-)

Is there a specific reason that `match` is obtained in the above way?

Reply:
Following up on this: The failing tests appear when trying to match `Defined_Binary_Op` and `Defined_Unary_Op` instead of one of the subclasses implementing these operator classes. If I understand correctly, the idea is that these should not actually be able to match successfully but only via their subclasses. 
Above changes in utils.py yield a match method for both classes (from STRINGBase) with wrong number of parameters (expects `pattern` as additional argument). Tests can be repaired when setting `match=None` in both classes, which could be a viable way of indicating "This class should not match".

Reply:
Thanks for all this investigation. When I suggested the change it was meant to make things simpler, not lead you into a warren of problems :-). You are, of course, right about not being able to do what I suggested for a static method - sorry about that.

In answer to your question about getting the match method using __dict__, this part of the code is inherited and not previously examined by me, so I don't know why it was implemented this way. I can't see any reason why it should stay as it is and I prefer the ability for subclasses to override match methods so would be very happy for you to make the change to use `hasattr` and `getattr` if you don't mind.

As for the failing test, I don't think the two offending classes should inherit from STRINGBase as they do not use this functionality. If they are changed to inherit from Base instead then the tests are OK (as it has no match function). If you are also happy to make that change as well, that would be great.

Also, many thanks for finding the very nice `classmethod` solution, it could be very useful as we inherit our way towards Fortran2018!

Reply:
Many thanks for your reply! I hope you don't mind me nudging.

Your suggestion sounds great, I'll happily do the changes. Would you like to have them as a separate PR or is it fine in this one?

Reply:
The answer was this PR please, but you've already done that :-).